k2/csrc/host/intersect.cc

```diff
@@ -91,35 +89,36 @@ void Intersection::GetSizes(Array2Size<int32_t> *fsa_size) {
     auto b_arc_iter_begin = arc_b_begin + b_.indexes[state_b];
     auto b_arc_iter_end = arc_b_begin + b_.indexes[state_b + 1];
 
-    // As both `a` and `b` are arc-sorted, we first process epsilon arcs.
-    // Noted that at most one for-loop below will really run as either `a` or
-    // `b` is epsilon-free.
-    for (; a_arc_iter_begin != a_arc_iter_end; ++a_arc_iter_begin) {
-      if (kEpsilon != a_arc_iter_begin->label) break;
-
-      StatePair new_state{a_arc_iter_begin->dest_state, state_b};
-      int32_t new_state_index = InsertIntersectionState(
-          new_state, &state_index_c, &qstates, &state_pair_map);
-      arcs_.emplace_back(curr_state_index, new_state_index, kEpsilon,
-                         a_arc_iter_begin->weight);
-      arc_map_a_.push_back(
-          static_cast<int32_t>(a_arc_iter_begin - arc_a_begin));
-      arc_map_b_.push_back(arc_map_none);
-    }
-    for (; b_arc_iter_begin != b_arc_iter_end; ++b_arc_iter_begin) {
-      if (kEpsilon != b_arc_iter_begin->label) break;
-      StatePair new_state{state_a, b_arc_iter_begin->dest_state};
-      int32_t new_state_index = InsertIntersectionState(
-          new_state, &state_index_c, &qstates, &state_pair_map);
-      arcs_.emplace_back(curr_state_index, new_state_index, kEpsilon,
-                         b_arc_iter_begin->weight);
-      arc_map_a_.push_back(arc_map_none);
-      arc_map_b_.push_back(
-          static_cast<int32_t>(b_arc_iter_begin - arc_b_begin));
-    }
+    if (treat_epsilons_specially_) {
+      // As both `a` and `b` are arc-sorted, we first process epsilon arcs.
+      // Noted that at most one for-loop below will really run as either `a` or
+      // `b` is epsilon-free.
```

Comment:
@danpovey , if `treat_epsilons_specially_ = true`, I think we should make sure here that at least one of `a` and `b` is epsilon-free.

Reply:
I decided to remove that check because there are situations where it produces the correct output, e.g. with max semiring you don't mind the extra paths with different orderings of epsilons.